tests/test_downloadermiddleware.py

```diff
@@ -71,7 +71,7 @@ def test_3xx_and_invalid_gzipped_body_must_redirect(self):
         In particular when some website returns a 30x response with header
         'Content-Encoding: gzip' giving as result the error below:
 
-            exceptions.IOError: Not a gzipped file
+            BadGzipFile: Not a gzipped file (...)
```

Comment:
To reproduce it, you can do:

```python
import gzip
import io

buff = io.BytesIO(b'aaaa') 
buff.seek(0) 
gz = gzip.GzipFile(fileobj=buff) 
gz.read()

Reply:
Maybe the test should assert that `BadGzipFile` is raised, then?

It may make sense to add a comment which links this to the redirect middleware `EXCEPTIONS_TO_RETRY` attribute, and test that `BadGzipFile` is a subclass of `OSError`, just in case that changes in some future Python version, and we need to update `EXCEPTIONS_TO_RETRY` accordingly.

Reply:
Well, this test actually tests that the default middleware order prevents the httpcompression middleware from seeing this response.

Not sure if we have a test that a similar 200 response is handled by retrying via the httpcompression and retry middlewares working together.

Reply:
`test_200_and_invalid_gzipped_body_must_fail` (which is the next test here) handles the second case, and while the code raises `BadGzipFile` the tests expects `IOError` so I think we have everything we need, except maybe better comments.